distribution/binary/compiler.bzl

```diff
@@ -0,0 +1,116 @@
+# Compile targets using a bazel-defined compilation mode.
+#
+# This is useful for building packages comprised of binaries built in
+# multiple modes - eg a release containing both `opt` and `dbg` builds.
+#
+# In this case the cli-supplied `-c ...` flag is ignored.
+#
+# NB: Avoid use of this as it creates a large sub-graph independent
+#  of the normal bazel build graph.
+#
+#  See https://bazel.build/extending/config#memory-performance-considerations
+#
+# The `bundled` rule allows you to define a set of compilation targets that you wish to
+# be compiled with compilation settings defined in bazel, that would normally be
+# set on the command line.
+#
+# The `mode` sets the `compilation_mode` - one of `dbg`, `opt`, or `fastbuild`.
+#
+# You can define the `targets` to be compiled, along with the destination `path`
+# of the compiled binary.
+#
+# You can also define lists of `stripopts` and `linkopts`.
+#
+# For example, you can bundle some stripped binaries together in `opt` mode:
+#
+# ```starlark
+#
+# bundled(
+#     name = "envoy",
+#     mode = "opt",
+#     stripopts = ["--strip-all"],
+#     targets = {
+#         "//distribution:envoy-binary": "envoy",
+#         "//distribution:envoy-contrib-binary": "envoy-contrib",
+#         "//external:su-exec": "utils/su-exec",
+#     },
+# )
+#
+#
+# The bundled output can then be packaged together with other, differently compiled, output:
+#
+# ```starlark
+#
+# bundled(
+#     name = "envoy-debug",
+#     mode = "dbg",
+#     targets = {
+#         "//distribution:envoy-binary": "envoy",
+#         "//distribution:envoy-dwarf": "envoy.dwp",
+#         "//distribution:envoy-contrib-binary": "envoy-contrib",
+#         "//distribution:envoy-contrib-dwarf": "envoy-contrib.dwp",
+#     },
+# )
+#
+# ```
+#
+
+def _compilation_config_impl(settings, attr):
+    _ignore = settings
+    return {
+        "//command_line_option:compilation_mode": attr.mode,
+        "//command_line_option:linkopt": attr.linkopts,
+        "//command_line_option:stripopt": attr.stripopts,
+    }
+
+compilation_config = transition(
+    implementation = _compilation_config_impl,
+    inputs = [],
+    outputs = [
+        "//command_line_option:compilation_mode",
+        "//command_line_option:linkopt",
+        "//command_line_option:stripopt",
+    ],
+)
+
+def _bundled_impl(ctx, **kwargs):
+    output_files = []
+
+    for i, target in enumerate(ctx.files.targets):
+        path = ctx.attr.targets.values()[i]
+        output_file = ctx.actions.declare_file(
+            "%s/%s/%s" %
+            (
+                ctx.attr.name,
+                ctx.var["COMPILATION_MODE"],
+                path,
+            ),
+        )
+
+        ctx.actions.run(
```

Comment:
instead of actually copying should we use `ctx.actions.symlink`?

Reply:
i wondered about this - i wasnt sure if it would work as expected - but i think getting that working would be worth the effort the files are huge

ill reWIP whille i test this further

Reply:
yea i would assume the next step would just be to `cp` the files out somewhere else, which seems like it should work fine. i imagine this might work more easily on windows too? 🤔

Reply:
stream to a destination directory? that is the bit i havent figured out - at least without a bit of path mangiling